Add c-shared and c-archive link modes

steeve opened the pull request:
This PR does the following:
- Add `c-shared` and `c-archive` link modes
- Exports a `<name>.cc` target for rapid inclusion as dep for `cc_` rules
- Changes the link builder to know the buildmode and output files
- When using the `c-archive` mode, strips the non-deterministic data from the archive from inside the builder
  - This could be reused in the pack builder, not early check suggest `go tool pack` is already deterministic (I may be wrong)
  - I could have went through and did it anyway, but it would require factoring `pack` a bit to clean the duplicates with `ar.go`

steeve commented:
Hum looking at the code in `ar.go` I need to clean the error handling a bit, but the gist should be there.

jayconrod commented on go/private/rules/wrappers.bzl:
```diff
@@ -61,6 +66,45 @@ def go_binary_macro(name, **kwargs):
   go_binary(name = name, **kwargs)
+  _go_binary_c_archive_shared(name, kwargs)
+
+def _go_binary_c_archive_shared(name, kwargs):
```
FYI, again to echo my comment about the C++ sandwich above, I think this interface is fine for now, but it may need to change in the future.

steeve replied:
i'd love to! but for now it'll have to do i think

jayconrod commented on go/tools/builders/ar.go:
```diff
@@ -0,0 +1,133 @@
+	EndRaw      [2]byte
+}
+
+func (h *header) Name() string {
```
`Name` doesn't seem to be called. Better remove it, since it doesn't handle long names.

steeve replied:
i left it because i thought `pack.go` could use it, but not feeling strongly about this

jayconrod commented on go/tools/builders/ar.go:
```diff
@@ -0,0 +1,133 @@
+	return strings.TrimRight(string(h.NameRaw[:]), " ")
+}
+
+func (h *header) Size() int64 {
```
Nit: I don't think `Size`, `Next` or `Deterministic` need to be exported. I guess the fields need to be public though, since `binary` uses reflection.

steeve commented on go/private/mode.bzl:
```diff
@@ -78,6 +82,8 @@ def get_mode(ctx, host_only, go_toolchain, go_context_data):
       "pure" in ctx.features,
   )
   linkmode = getattr(ctx.attr, "linkmode", LINKMODE_NORMAL)
+  if linkmode in [LINKMODE_C_SHARED, LINKMODE_C_ARCHIVE]:
```
Sure thing

steeve commented on go/private/rules/binary.bzl:
```diff
@@ -66,10 +66,15 @@ def _go_binary_impl(ctx):
   return [
       library, source, archive,
+      OutputGroupInfo(
+          binary = outs,
```
actually I think I made a typo. I wanted to separate the binary from its headers, but `cgo_exports` is already defined so...

steeve replied:
The issue is that in the case of shared or archive, the binary is not an executable, actually. Perhaps I should remove it from executable in that case ?

steeve commented on go/tools/builders/ar.go:
```diff
@@ -0,0 +1,133 @@
+// - File Modes
+// The archive is modified in place.
+func stripArMetadata(archivePath string) error {
+	stat, err := os.Stat(archivePath)
```
Indeed, but I used it to get the current `FileMode`

jayconrod replied:
You can pass 0 for the mode though. It's only used when creating a new file.

steeve commented on go/tools/builders/ar.go:
```diff
@@ -0,0 +1,133 @@
+	defer archive.Close()
+
+	magic := make([]byte, len(arHeader))
+	if n, err := archive.Read(magic); err != nil {
```
sure thing

steeve commented on go/tools/builders/ar.go:
```diff
@@ -0,0 +1,133 @@
+	copy(h2.ModTimeRaw[:], zeroBytes)
+	copy(h2.OwnerIdRaw[:], zeroBytes)
+	copy(h2.GroupIdRaw[:], zeroBytes)
+	copy(h2.FileModeRaw[:], zeroBytes)
```
I was with your on this, but then I check what gnu `ar` does in deterministic mode (https://sourceware.org/binutils/docs-2.26/binutils/ar-cmdline.html), and it also clears the mode

jayconrod replied:
Got it. Surprising, but seems to be the correct thing to do.

steeve commented:
Also I was thinking about maybe not include the linkmode in case of c-archive and c-shared so as to be bootstrap the stdlib multiple times (which only makes sense in the `shared` and `plugin` modes). What do you think?

steeve commented:
Added tests and fixed a bug in `c-shared` mode that would break proper inclusion: dynamic libraries have to be named `libXXX`.

Run tests with:
```
$ bazel test //tests/core/c_shared:all //tests/core/c_archive:all
```

steeve commented:
Also, the exported `.cc` target is now a `cc_library` instead of a `cc_import` in order to make the header easily accessible and have better support for inclusion across the board (`java_library` is picky for instance).

steeve commented:
I could not reproduce the test failures on the CI...

steeve commented:
Refactored the tests to have them in the same package.
Test now with:
```
$ bazel test //tests/core/c_linkmodes:all
```

jayconrod commented:
Went a bit deeper today. I was getting undefined reference errors in my build script because I was neglecting to pack the C objects into the Go archive. Easily fixed.

The main obstacle now is that we need to compile the standard library in shared mode and link against that. I updated my [build script](https://gist.github.com/jayconrod/462442345eb70c0f806e8dde11086dcc) to do this, but I'm still getting an error when linking the .so file.

```
+ go tool link -o /usr/local/google/home/jayconrod/go/src/example.com/repo/_obj/repo/repo.so -installsuffix=shared -buildmode=c-shared -linkmode=external -extld=gcc /usr/local/google/home/jayconrod/go/src/example.com/repo/_obj/repo/_pkg_.a
/usr/local/google/home/jayconrod/go/src/example.com/repo/_obj/stdlib/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: /tmp/go-link-459261086/go.o: relocation R_X86_64_TPOFF32 against `runtime.tlsg' can not be used when making a shared object; recompile with -fPIC
```

steeve commented on go/private/actions/binary.bzl:
```diff
@@ -24,7 +34,13 @@ def emit_binary(go,
-  executable = go.declare_file(go, name=name, ext=go.exe_extension)
+  extension = go.exe_extension
+  if go.mode.link == LINKMODE_C_SHARED:
+    name = "lib" + name # shared libraries need a "lib" prefix in their name
```
Apparently when linking against dynamic libraries with `cc_` rules, the `lib` prefix is mandatory. Same goes for Java's `System.loadLibrary`.
See https://docs.bazel.build/versions/master/be/c-cpp.html#cc_library.linkshared
> Create a shared library. To enable this attribute, include linkshared=1 in your rule. By default this option is off. If you enable it, you must name your binary libfoo.so (or whatever is the naming convention of libraries on the target platform) for some sensible value of foo.

I'm fine with a custom name though, but then we need to ensure the name is unique to the `WORKSPACE` then?

steeve commented:
I got to ask, why do you want to link the library in shared mode? I'm not doing it at the moment and it's working fine? Is that in case multiple go binaries are loaded in the same process?

For the record, there is code in https://github.com/golang/mobile/blob/master/cmd/gomobile/bind_androidapp.go#L24 that references https://github.com/golang/go/issues/13234 which seems to suggest this is not possible to link the `stdlib` in `c-shared` mode.

`golang/mobile` does the following:
```go
	androidArgs := []string{"-gcflags=-shared", "-ldflags=-shared"}
```

steeve commented:
For the record I've done some small modifications/improvements, however, the tests indeed fail on linux, for either:
Missing `pthread` symbols:
```
$ bazel test //tests/core/c_linkmodes:all
Loading:
Loading: 0 packages loaded
Analyzing: 30 targets (15 packages loaded)
INFO: Analysed 30 targets (16 packages loaded).
INFO: Found 28 targets and 2 test targets...
[2 / 11] [-----] BazelWorkspaceStatusAction stable-status.txt
ERROR: /Users/steeve/go/src/github.com/znly/rules_go/tests/core/c_linkmodes/BUILD.bazel:14:1: Linking of rule '//tests/core/c_linkmodes:c-archive_test' failed (Exit 1)
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a(000006.o):gcc_libinit.c:function _cgo_try_pthread_create: error: undefined reference to 'pthread_create'
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a(000006.o):gcc_libinit.c:function _cgo_try_pthread_create: error: undefined reference to 'pthread_detach'
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a(000007.o):gcc_linux_amd64.c:function x_cgo_init: error: undefined reference to 'pthread_attr_getstacksize'
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a(000007.o):gcc_linux_amd64.c:function _cgo_sys_thread_start: error: undefined reference to 'pthread_sigmask'
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a(000007.o):gcc_linux_amd64.c:function _cgo_sys_thread_start: error: undefined reference to 'pthread_attr_getstacksize'
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a(000007.o):gcc_linux_amd64.c:function _cgo_sys_thread_start: error: undefined reference to 'pthread_sigmask'
collect2: error: ld returned 1 exit status
INFO: Elapsed time: 1.916s, Critical Path: 0.13s
FAILED: Build did NOT complete successfully
//tests/core/c_linkmodes:c-shared_test                                NO STATUS
```

Or at linking:
```
$ bazel test //tests/core/c_linkmodes:all
Loading:
Loading: 0 packages loaded
INFO: Analysed 30 targets (16 packages loaded).
Building: no action
INFO: Found 28 targets and 2 test targets...
[0 / 9] [-----] BazelWorkspaceStatusAction stable-status.txt
ERROR: /Users/steeve/go/src/github.com/znly/rules_go/tests/core/c_linkmodes/BUILD.bazel:20:1: GoLink tests/core/c_linkmodes/linux_amd64_stripped_c-shared/libadder_shared.so failed (Exit 1)
/root/.cache/bazel/_bazel_root/e865bbc31a8eba0f1feabab3967a409e/bazel-sandbox/888423118525108281/execroot/io_bazel_rules_go/bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-shared/stdlib~/pkg/tool/linux_amd64/link: running /usr/bin/gcc failed: exit status 1
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
/usr/bin/ld.gold: error: /tmp/go-link-173844341/go.o: unsupported reloc 23 against local symbol
```

steeve commented:
Ok got it, it is because of the `-shared` parameter needs to be passed to all rules.
I'll submit the fix later tonight.

jayconrod commented:
I'm still trying to figure out to build the standard library in shared mode. `-gcflags=shared -ldflags=shared` doesn't seem to do the trick. The two people I'm sure would know are either travelling or on parental leave right now. I've emailed golang-dev, hopefully we can get an answer there.

steeve commented:
According to https://groups.google.com/d/msg/golang-nuts/EhndTzcPJxQ/0opO-W7TBwAJ:
> You should try building your Go code with the Go equivalent of -fPIC,  which is `go build -buildmode=c-archive -gcflags=-shared`

I've added it to `stdlib` (https://github.com/bazelbuild/rules_go/pull/1365/commits/33fa2593d2e19d6e24a902b7f90a2692ef30a833#diff-0d7f9c8aa3da936c11967a5fded97d9fR72), `compile` and `asm`

steeve commented:
I should've told that I have the tests passing on Linux as a result.

jayconrod commented:
Thanks for linking that thread. The flag I was missing when building the standard library was `-asmflags=shared`.

There are still some tests failing, but I think we're very close. I'll block out some time tomorrow to go through this.

steeve commented:
Fixed the tests. One was me missing to forward the `extldflags` properly, the other had to do with adding `-trimpath` as a `gcflags` when go installing stdlib. Somehow removing it fixed the test, so here goes.

jayconrod commented:
FYI, here's the thread I started on golang-dev about building the standard library in shared mode. https://groups.google.com/forum/#!topic/golang-dev/MZ3_UobYoFc

steeve commented:
Good to know for Go 1.10, although the weird thing is that I'm not doing it yet tests passes?
I can definitely add a check for Go 1.10 to add =all

steeve commented:
Ok say the `=all` thingy is proving more complex that I thought. That flag appears at go1.10, so I need to check wether I'm go 1.10 at runtime. The only reliable way to do that is via build tags (the `go1.10` will set onward, as all past versions).

I make a rule to auto generate those files, so that there can be check at runtime with the following template:
```python
_GO_VERSION_CHECK_TPL = """
// +build go{0}.{1}

package main

const isGoVersion{0}_{1} = true
"""
```

Those files are attached to the builders.

steeve commented:
Hum it seems `go tool compile` is ignoring the build tags when bootstrapping..

steeve commented:
Actually, that may not be needed since we are building packages by packages, and std and runtime are built also by hand.

steeve commented:
I ultimately went for a much simpler solution 😄 !

jayconrod commented:
There are test failures on Travis CI due to `cc_import`, which was introduced after Bazel 0.8.0, which is our oldest supported version. That's quite old at this point, and I'm fine with dropping support in the next release. We'll probably need to do so soon anyway for other C++ issues.

#1384 will update Travis CI to test against Bazel 0.10.0.

steeve commented on go/private/rules/wrappers.bzl:
```diff
@@ -61,6 +66,56 @@ def go_binary_macro(name, **kwargs):
+    name = cc_library_name,
+    hdrs = [c_hdrs],
+    deps = [cc_import_name],
+    includes = ["."],
```
good call !

steeve replied:
How did you manage to trigger this error ?
I think I should create a directory and put the .h there, then ?

steeve replied:
Basically it boils down to how we want to include the .h

```
#include "my_go_lib.h" ?
#include "my_golib/my_go_lib.h" ?
```

steeve replied:
Fixed by https://github.com/bazelbuild/rules_go/pull/1365/commits/6e4fe8e228269df70df38ee908183cf8715a18ce

steeve commented on go/tools/builders/env.go:
```diff
@@ -49,6 +62,31 @@ func abs(path string) string {
 }
 
+func absoluteCgoFlags(flags []string) []string {
+	ret := flags[:]
```
good call !

steeve commented on go/tools/builders/env.go:
```diff
@@ -49,6 +62,31 @@ func abs(path string) string {
 	}
 }
 
+func absoluteCgoFlags(flags []string) []string {
```
This is needed in general because it seems a lot of CROSSTOOLs are built using relative paths, which go won't understand since it cd's in $WORK and runs commands from there...

About the flag list, indeed it's difficult to have a complete list, so I figure we'd just add them as needed later on. So far I haven't needed more (and I'm building for iOS and Android).

steeve replied:
Fixed in the new pushed version

steeve commented on go/tools/builders/stdlib.go:
```diff
@@ -59,10 +61,27 @@ func run(args []string) error {
+
+	allSlug := ""
+	if goReleaseTags["go1.10"] {
+		allSlug = "all="
```
sounds good

steeve commented on go/tools/builders/go_tags.go:
```diff
@@ -0,0 +1,40 @@
+
+	// goReleaseTags holds the release tags with which Go was built as a map.
+	// This is useful for detecting mininum required Go versions.
+	goReleaseTags = map[string]bool{}
```
I figure there might be some other checks in the future (such as enumerating them, for instance). I'll move all this to env.go though.

steeve commented on go/tools/builders/go_tags.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+var (
+	// goBuildTags holds the build tags with which Go was built as a map.
```
That's not really the case, since we now support the `gotags` attribute which controls this. Also, in the case of iOS, the "ios" tag is there, good.

steeve commented:
Put the shared flag in the environment since it needs to be passed to both `compile`, `stdlib` and `asm`.

jayconrod commented:
I spent a few hours this morning trying to figure out why that tests are failing. I think I understand now.

When we run `go tool cgo`, the contents of the `CGO_CPPFLAGS`, `CGO_CFLAGS`, `CGO_CXXFLAGS`, and `CGO_LDFLAGS` environment variables are combined with the `#cgo` directives in source files with the same name. These variables are used by cgo (it invokes the C compiler). The new part that I didn't know before is that the LDFLAGS are embedded in pragmas in the generated .go files. These look like `//go:cgo_ldflag ...`. The compiler copies these into generated .a files, and the linker reads those flags and passes them to the external linker.

The reason tests are failing is that we are absolutizing paths in `CGO_LDFLAGS`, specifically the path to a .so file in `cdeps`. The path where this .so is available is different in the CgoCodegen and GoLink actions because of sandboxing: the sandbox path has a random number as part of its name. This is why the test passes on macOS (which has sandboxing disabled by default) and Travis CI (which has sandboxing disabled explicitly). 

----

We've been iterating on the PR for a couple weeks now. In order to make progress, please break this up into smaller pieces.

1. This PR should reduced to just add support for c-archive and c-shared build modes and nothing more. Just the minimum needed to pass CI tests in the default environment for Linux and macOS.
2. Changes needed to support custom CROSSTOOL. This will be the path absolutization stuff. But follow the discussion in #1357. I'll post a comment there shortly with more analysis.
3. Additional declarations currently in wrappers.bzl. These are changes to the public interface, so I'd like to be careful here and consider these separately after (1) has gone in.

steeve commented:
The thing is I'm not able to reproduce the failing tests on Linux (ubuntu running in Docker).
Also Darwin does have sandboxing enabled by default.

steeve commented:
To answer your questions:
1. No worries. Although I do not reproduce the failures in CI on my Linux environment.
2. I'll sent it in a separate PR. However I don't get reproducibility issues with it. The ones I had were caused by sending `-trimpath $(pwd)` when `go install`ing `stdlib`, which was fixed.
3. I would completely agree with you if not for the fact that we need to expose the generated header too, which is why I put them in the interface to begin with. As you can see this is not trivial to consume so I don't think removing it from this PR is worth it.

Could you point me to the failing tests output? The ones I see in CI are related to a `PrintCXXVersion()` which I cannot reproduce and doesn't seem related to this PR ?

jayconrod commented:
About reproducing the failure, this is the output I'm seeing it buildkite:

```
/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/0cbda4e4f9fcbe25e901dd8b65edca92/bazel-sandbox/5239077463959946252/execroot/io_bazel_rules_go/bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/tool/linux_amd64/link: running /usr/bin/gcc failed: exit status 1
/usr/bin/ld.gold: error: cannot find -lc_version
bazel-out/k8-fastbuild/bin/tests/legacy/examples/cgo/cc_dependency/libversion.a(cxx_version.pic.o):cxx_version.cc:function PrintCXXVersion: error: undefined reference to 'PrintCVersion'
collect2: error: ld returned 1 exit status
```

The target that's failing is `//tests/legacy/examples/cgo:cgo_lib_test` and a couple others.

Bazel has different sandboxing behavior on different OSs. I think the random number in the sandbox directory is relatively recent. Maybe it behaves differently in a Docker container? Do you have access to a Linux VM or native Linux box?

If it would be easier, I can edit and test this PR directly. Let me know.

About (3), there are still some changes there I'd like. I think it would be easier to discuss these in a separate PR.

* `_go_binary_c_archive_shared` should probably be in cgo.bzl.
* Most of the rules should be `//visibility:private`.
* I want to be absolutely sure the `cc_library` rule is needed. I'm hoping that `cc_import` is sufficient.
* I want to be absolutely sure the `includes` attribute is needed. I'm hoping that `hdrs` is sufficient.

steeve commented:
Doing two consecutive runs in a Docker container and comparing the hashes gave me a list of affected packages, I think those are CGo packages.
```
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/sha512.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/net.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/poly1305.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/sha512.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/poly1305.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/os/user.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/sync/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/chacha20poly130
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/plugin.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime/internal/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/internal/cipherhw.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/os/signal.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/sha256.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/aes.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/internal/cipherhw.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/sha1.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime/debug.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime/cgo.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/aes.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/rc4.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/os/signal.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/os/user.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/math/big.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/chacha20poly1305.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/math/big.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/strings.a
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/math.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/sha256.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/reflect.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/internal/cpu.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/md5.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/elliptic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/rc4.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime/debug.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/net.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/syscall.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/sha1.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/syscall.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime/internal/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/plugin.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/os/signal/internal/pty.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime/cgo.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/sync/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/strings.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/hash/crc32.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/elliptic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/curve25519.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/hash/crc32.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/os/signal/internal/pty.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/math.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/curve25519.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/internal/cpu.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/reflect.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/md5.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/math.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/poly1305.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/net.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/sha1.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/elliptic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/sync/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/internal/cipherhw.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/elliptic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime/cgo.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/os/user.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/syscall.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/poly1305.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/math/big.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/os/signal/internal/pty.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime/cgo.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime/debug.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/reflect.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/md5.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/math.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime/internal/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/chacha20poly1305.a
bazel-out/k8-fastbuild/bin/tests/core/c_linkmodes/linux_amd64_stripped_c-archive/adder_archive.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/os/signal.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/internal/cpu.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/strings.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/sha512.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/curve25519.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/sync/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/sha256.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/sha512.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/plugin.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/rc4.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/syscall.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/runtime/internal/atomic.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/hash/crc32.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/md5.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/net.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime/debug.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/hash/crc32.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/aes.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/internal/cpu.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/curve25519.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/sha256.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/sha1.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/math/big.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/internal/cipherhw.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/runtime.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/vendor/golang_org/x/crypto/chacha20poly130
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/os/user.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/crypto/rc4.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/reflect.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/strings.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/plugin.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped_c-archive/stdlib~/pkg/linux_amd64/crypto/aes.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/os/signal.a
bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/pkg/linux_amd64/os/signal/internal/pty.a
```

steeve commented:
Looking at `signal.a`, here is the diff between two runs:
```diff
8,9c8,9
< 00000070  62 75 69 6c 64 20 69 64  20 22 49 69 78 62 6e 54  |build id "IixbnT|
< 00000080  46 35 61 53 30 4e 61 59  50 45 49 53 78 73 2f 73  |F5aS0NaYPEISxs/s|
---
> 00000070  62 75 69 6c 64 20 69 64  20 22 58 59 7a 63 77 6e  |build id "XYzcwn|
> 00000080  63 4e 4f 4b 4e 59 7a 35  57 53 44 70 43 44 2f 73  |cNOKNYz5WSDpCD/s|
12,13c12,13
< 000000b0  20 69 64 20 22 49 69 78  62 6e 54 46 35 61 53 30  | id "IixbnTF5aS0|
< 000000c0  4e 61 59 50 45 49 53 78  73 2f 73 71 38 38 4e 51  |NaYPEISxs/sq88NQ|
---
> 000000b0  20 69 64 20 22 58 59 7a  63 77 6e 63 4e 4f 4b 4e  | id "XYzcwncNOKN|
> 000000c0  59 7a 35 57 53 44 70 43  44 2f 73 71 38 38 4e 51  |Yz5WSDpCD/sq88NQ|
37,38c37,38
< 00000240  49 69 78 62 6e 54 46 35  61 53 30 4e 61 59 50 45  |IixbnTF5aS0NaYPE|
< 00000250  49 53 78 73 2f 73 71 38  38 4e 51 53 47 58 6a 6b  |ISxs/sq88NQSGXjk|
---
> 00000240  58 59 7a 63 77 6e 63 4e  4f 4b 4e 59 7a 35 57 53  |XYzcwncNOKNYz5WS|
> 00000250  44 70 43 44 2f 73 71 38  38 4e 51 53 47 58 6a 6b  |DpCD/sq88NQSGXjk|
```

It seems this is because of build id.

steeve commented:
I could also reproduce it by disabling `absoluteCgoArgs`, see one example of two consecutive runs with `bazel clean --expunge` between them:
```
$ diff signal.a.3.hex signal.a.4.hex
8,9c8,9
< 00000070  62 75 69 6c 64 20 69 64  20 22 58 5f 4f 4b 64 43  |build id "X_OKdC|
< 00000080  33 72 61 59 33 45 7a 46  34 35 4c 41 48 63 2f 73  |3raY3EzF45LAHc/s|
---
> 00000070  62 75 69 6c 64 20 69 64  20 22 73 51 54 72 52 6b  |build id "sQTrRk|
> 00000080  45 43 53 4c 45 32 66 6a  31 33 58 50 33 66 2f 73  |ECSLE2fj13XP3f/s|
12,13c12,13
< 000000b0  20 69 64 20 22 58 5f 4f  4b 64 43 33 72 61 59 33  | id "X_OKdC3raY3|
< 000000c0  45 7a 46 34 35 4c 41 48  63 2f 73 71 38 38 4e 51  |EzF45LAHc/sq88NQ|
---
> 000000b0  20 69 64 20 22 73 51 54  72 52 6b 45 43 53 4c 45  | id "sQTrRkECSLE|
> 000000c0  32 66 6a 31 33 58 50 33  66 2f 73 71 38 38 4e 51  |2fj13XP3f/sq88NQ|
37,38c37,38
< 00000240  58 5f 4f 4b 64 43 33 72  61 59 33 45 7a 46 34 35  |X_OKdC3raY3EzF45|
< 00000250  4c 41 48 63 2f 73 71 38  38 4e 51 53 47 58 6a 6b  |LAHc/sq88NQSGXjk|
---
> 00000240  73 51 54 72 52 6b 45 43  53 4c 45 32 66 6a 31 33  |sQTrRkECSLE2fj13|
> 00000250  58 50 33 66 2f 73 71 38  38 4e 51 53 47 58 6a 6b  |XP3f/sq88NQSGXjk|
```

steeve commented:
I think the key is in `go tool compile`'s `-buildid` flag maybe ?

steeve commented:
I could finally reproduce the failures in CI in my own Docker environment. I've remove the absolute args for now as it's not blocking for this PR.

I've also did what you suggested.

About the `cc_library`, I would keep it because of rules only take it as input and don't work with `cc_import` (most notably `java_library`), and I think it doesn't hurt.
As for `includes`, well if we drop it, it means we'll have to include the cgo_export via it's full workspace path (which may not be a bad thing?). That's all there is to it really.

steeve commented:
For instance, the tests goes from:
```c
#include "adder_archive.h"
```
to
```c
#include "tests/core/c_linkmodes/adder_archive.h"
```

Thinking about it, I find the later nicer.

jayconrod commented:
All tests passing, merged.

Thanks so much for submitting this! It's a really valuable feature to have, and this gets us a long way toward support for plugins, too. And thanks for being patient with all my code review comments.
